examples/hitl/rearrange_v2/app_state_load_episode.py

```diff
@@ -0,0 +1,125 @@
+
+# Copyright (c) Meta Platforms, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+from app_data import AppData
+from app_state_base import AppStateBase
+from app_states import (
+    create_app_state_cancel_session,
+    create_app_state_end_session,
+    create_app_state_start_screen,
+)
+from session import Session
+from util import get_top_down_view
+
+from habitat_hitl.app_states.app_service import AppService
+from habitat_hitl.core.user_mask import Mask
+
+
+class AppStateLoadEpisode(AppStateBase):
+    """
+    Load an episode.
+    A loading screen is shown while the content loads.
+    * If a next episode exists, launch RearrangeV2.
+    * If all episodes are done, end session.
+    * If any user disconnects, cancel the session.
+    """
+
+    def __init__(
+        self, app_service: AppService, app_data: AppData, session: Session
+    ):
+        super().__init__(app_service, app_data)
+        self._session = session
+        self._loading = True
+        self._session_ended = False
+        self._frame_number = 0
+        self._save_keyframes = False
+
+    def get_next_state(self) -> Optional[AppStateBase]:
+        if self._cancel:
+            return create_app_state_cancel_session(
+                self._app_service,
+                self._app_data,
+                self._session,
+                "User disconnected.",
+            )
+        if self._session_ended:
+            return create_app_state_end_session(
+                self._app_service, self._app_data, self._session
+            )
+        # When all clients finish loading, show the start screen.
+        if not self._loading:
+            return create_app_state_start_screen(
+                self._app_service, self._app_data, self._session
+            )
+        return None
+
+    def sim_update(self, dt: float, post_sim_update_dict):
+        self._status_message("Loading...")
+
+        # Skip a frame so that the status message reaches the client before the server loads the scene and blocks.
+        if self._frame_number == 1:
+            self._increment_episode()
+        # Once the scene loaded, show a top-down view.
+        elif self._frame_number > 1:
+            cam_matrix = get_top_down_view(self._app_service.sim)
+            post_sim_update_dict["cam_transform"] = cam_matrix
+            self._app_service._client_message_manager.update_camera_transform(
+                cam_matrix, destination_mask=Mask.ALL
+            )
+            # Wait for clients to signal that content finished loading on their end.
+            # HACK: The server isn't immediately aware that clients are loading. For now, we simply skip some frames.
+            # TODO: Use the keyframe ID from 'ClientMessageManager.set_server_keyframe_id()' to find the when the loading state is up-to-date.
+            if self._frame_number > 20:
```

Comment:
This hack will be addressed in a future PR.

The correct approach to handle this is to look at the frame number when the new scene keyframe is sent, and wait for all clients to echo back this frame number.

This feature is supported but coupled with other systems. See `ClientHelper`.

Reply:
This is to ensure that the server doesn't progress to the start screen until all clients finished downloading the content.